### DISCUSSION TITLE
What should I use for node __dirname in TypeScript?

### DISCUSSION THREAD

**Discussion by aderchox** [egoist/tsup#708] [Q&A]
In Node, MJS has `url.fileURLToPath(import.meta.url)` and CJS has `__dirname`, so what should I use for TypeScript?
I'm sort of using \*both\* module formats right now, because I'm using tsup to convert to both formats, so there must be something that translates to both of them? Or maybe I should use only one of them, in that case, is there a preference? Thank you in advance 🙏

**Comment by aderchox** [comment]
I've found a solution [on antfu's blog](https://antfu.me/posts/isomorphic-dirname) but it's still problematic when running tests using jest (ts-jest):
> error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.

I have tried this in my tsconfig.json:
```js
{
  "compilerOptions": {
    "module": "ES2022", // tried NodeNext too
    "moduleResolution": "Node"
  },
}
```
But still no success, a similar question has also been asked on SO:
https://stackoverflow.com/questions/70440505/ts1343-the-import-meta-meta-property-is-only-allowed-when-the-module-opti
And an apparently related issue on GH:
https://github.com/facebook/jest/issues/12952
Please let me know if there are any workarounds. Thank you.

**Comment by rxliuli** [comment]
Whichever you use, they will be properly escaped as long as you add the `--shims` option
ref: <https://tsup-git-fork-rxliuli-dev-egoist.vercel.app/#inject-cjs-and-esm-shims>
